### PR TITLE
Fix tracking misattribution bug

### PR DESCRIPTION
**Pull request by ghostwords**
Reloading a tab that contains localStorage tracking, such as from an embedded YouTube video, while the tab is still loading runs into a Badger bookkeeping bug which causes localStorage tracking to be attributed to all frame 0 resources (instead of the actual frame localStorage was in).

- [x] Refactor [supercookie detection](https://github.com/EFForg/privacybadger/commit/8935682cb31be75f0c9b5d04d8ddd57297d950a1)
- [x] Remove fallback to frame 0 data when we don't have data for frame XX
- [x] Add test that catches misattribution bug
- [ ] Since this will only resolve the problem going forward, to fix overblocking for existing users, we should "yellowlist" affected domains (like `kxcdn.com`) or, more to the point, make Badger do a one-time cleanup for a set of false positives

### CODE REVIEW

**Comment by ghostwords** [comment]
@gunesacar Could you review my supercookie detection refactor here? I'm under the impression we can avoid a bunch of complexity by treating supercookie tracking the same way we treat canvas fingerprinting, that is by recording tracking whenever we get notified by the content script, entirely apart from the webRequest life cycle.

**Comment by gunesacar** [comment]
@ghostwords this looks good to me. Thanks for fixing this nasty bug!

**Comment by ghostwords** [comment]
Thanks for checking and thanks again for your help!

**Comment by ghostwords** [comment]
@cowlicks, could you review these changes too please? I should have the test ready soon. I'll look through some broken site reports afterwards, to build a list of domains to migrate. I don't expect any other changes here at this point though.

File: src/js/webrequest.js [hunk 1]
```diff
@@ -465,16 +448,15 @@ function recordFingerprinting(tabId, msg) {
 /**
  * Read the frame data from memory
  *
- * @param tabId TabId to check for
- * @param frameId FrameID to check for
+ * @param tab_id Tab ID to check for
```
Comment (cowlicks) [inline src/js/webrequest.js]: I prefer `lowercase_and_underscores` over `mixedCase` names *so much*, but it is inconsistent with the rest of the code.

Comment (ghostwords) [inline src/js/webrequest.js]: Agreed! I'm transitioning gradually to what we prefer.

Comment (ghostwords) [inline src/js/webrequest.js]: To be clear, I like Python style for non-objects and Java style for objects but I'm willing to simplify.

Once we have more consistent rules, we could update our ESLint config to enforce style stuff, if we wish.

File: src/js/background.js [hunk 2]
```diff
@@ -718,10 +713,6 @@ Badger.prototype = {
     chrome.browserAction.setTitle({tabId: tab.id, title: "Privacy Badger"});
   },
 
-  getFrameData: function(tabId, frameId){
-    return webrequest.getFrameData(tabId, frameId);
```
Comment (cowlicks) [inline src/js/background.js]: I think we can remove the `getFrameData` export from `webrequests.js` now.

File: src/js/webrequest.js (hunk 1 above)
Comment (cowlicks) [inline src/js/webrequest.js]: I'm fine with this as long as we are internally consistent moving forward.

**Comment by cowlicks** [comment]
I haven't removed the stuff related to storage yet. I'll follow up on that tomorrow.

**Comment by cowlicks** [comment]
I made a separate issue for some of the cleanups in `webrequests.js` https://github.com/EFForg/privacybadger/issues/1408 

Feel free to close it if you want to include the fixes here.

**Comment by ghostwords** [comment]
>All these references to tabData are surely causing our problems with race conditions.

Could you explain what you mean?

**Comment by ghostwords** [comment]
I'm going to rewrite this branch's history later to confirm the test fails prior to the fixes.

**Comment by cowlicks** [comment]
@ghostwords about `badger.tabData`: we're using it to store and share information about tabs and frames. And we are doing all of this asynchronous stuff in webrequests that reads and writes to `tabData` without any guarantee of thread-safety. For example [here](https://github.com/EFForg/privacybadger/blob/5a3f4475204033c0f25ccdd89100c4262e19e474/src/js/webrequest.js#L377-L389) in `webrequest.recordFingerPrinting`, we call `getFrameUrl`, which calls `getFrameData` which checks `badger.tabData` for the URL. But between the time the request triggered `recordFingerPrinting`, the tab URL in `tabData` could have changed, or the tab could have been removed from `tabData`. This could possibly cause us to misattribute the fingerprinting to the wrong origin, or no origin at all.

To fix this, each webrequest action should be as [atomic](https://en.wikipedia.org/wiki/Linearizability) as possible, and depend on as little shared state as possible. Some shared state is unavoidable, like the score against a tracker. 

But there are plenty of low hanging fruit here. We could easily avoid accessing `tabData` in `recordSuperCookie` and `recordFingerPrinting` since the information we want (the url and frame id) are already in the `details` object that we get for every request. Once we do this, we could remove the `getFrameData` and `getFrameUrl` functions completely.

**Comment by ghostwords** [comment]
What do you mean by storage-related changes?

**Comment by ghostwords** [comment]
OK, done rewriting. I had to make some fixes to the test. Could you look these changes over again?

I don't see what you mean about `tabData` changes (if that's what you were talking about) here being unrelated. Not returning frame 0 data as fallback also fixes the bug (see the Travis build results above).

Good points about `tabData`. We should investigate further, could you open an issue?

**Comment by cowlicks** [comment]
@ghostwords update: I'm currently trying to verify the new unittest fails on master

**Comment by cowlicks** [comment]
I've been running this script, which I modified by adding back `has_supercookie` tests to look  for this bug on master, but it isn't being triggered: https://gist.github.com/cowlicks/46382e9b3f20e7d564963f6c188169b2

**Comment by ghostwords** [comment]
Are you talking about the functional test here? You can just revert the last three commits containing fixes and see the test fail afterwards:
```bash
$ git checkout master
$ git merge fix-misattribution
$ git revert 30f3857 4911760 4692737
$ echo "Run supercookie tests" && make travisbuild && (cd tests && BROWSER=chrome ENABLE_XVFB=1 PB_EXT_PATH=../privacy_badger-2017.5.9.crx py.test -s -v --durations=10 selenium/super_cookie_test.py::SuperCookieTest) # this is in a virtualenv with all deps installed
...
================================================ FAILURES ================================================
______________________ SuperCookieTest.test_async_navigation_tracker_misattribution ______________________
...
E       AssertionError: Image is not a tracker but was flagged as one.
```

**Comment by ghostwords** [comment]
Re "isn't being triggered": As we discussed on yesterday's call, this bug doesn't populate `supercookie_domains`, it just updates `snitch_map` for the affected domains.

**Pull request merged**